Compiler stack overflow in cyclic base lookup (DIM)

The following test case crashes after https://github.com/dotnet/roslyn/pull/38564 has been merged

``` c#
        [Fact]
        public void NestedTypes_33()
        {
            var source1 =
@"
interface I1 : IA<int>.NF { }

interface IQ<T> { }

interface IA<T> : IB<IQ<T>> { }

interface IB<T> : IA<IQ<T>> { }
";

            var compilation1 = CreateCompilation(source1, options: TestOptions.DebugDll,
                                                 parseOptions: TestOptions.Regular);

            compilation1.VerifyDiagnostics(
                // some errors are expected:
                // error CS0426: The type name 'NF' does not exist in the type 'IA<int>'
                // error CS0529: Inherited interface 'IB<IQ<T>>' causes a cycle in the interface hierarchy of 'IA<T>'
                // error CS0529: Inherited interface 'IA<IQ<T>>' causes a cycle in the interface hierarchy of 'IB<T>'
                );
            CompileAndVerify(compilation1);
        }
```

I suspect the problem is a missing check `BaseTypeAnalysis.TypeDependsOn(...)` in the name lookup code added in #38564.

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
@@ -951,6 +951,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                 return ImmutableArray<NamedTypeSymbol>.Empty;
             }
 
+            var cycleGuard = ImmutableHashSet.Create(type.OriginalDefinition);
+
             // Consumers of the result depend on the sorting performed by AllInterfacesWithDefinitionUseSiteDiagnostics.
             // Let's use similar sort algorithm.
             var result = ArrayBuilder<NamedTypeSymbol>.GetInstance();
@@ -958,7 +960,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             for (int i = interfaces.Length - 1; i >= 0; i--)
             {
-                addAllInterfaces(interfaces[i], visited, result, basesBeingResolved);
+                addAllInterfaces(interfaces[i], visited, result, basesBeingResolved, cycleGuard);
             }
 
             result.ReverseContents();
@@ -970,17 +972,22 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             return result.ToImmutableAndFree();
 
-            static void addAllInterfaces(NamedTypeSymbol @interface, HashSet<NamedTypeSymbol> visited, ArrayBuilder<NamedTypeSymbol> result, ConsList<TypeSymbol> basesBeingResolved)
+            static void addAllInterfaces(NamedTypeSymbol @interface, HashSet<NamedTypeSymbol> visited, ArrayBuilder<NamedTypeSymbol> result, ConsList<TypeSymbol> basesBeingResolved, ImmutableHashSet<NamedTypeSymbol> cycleGuard)
             {
-                if (@interface.IsInterface && visited.Add(@interface))
+                if (@interface.IsInterface && !cycleGuard.Contains(@interface.OriginalDefinition) && visited.Add(@interface))
                 {
                     if (!basesBeingResolved.ContainsReference(@interface.OriginalDefinition))
                     {
                         ImmutableArray<NamedTypeSymbol> baseInterfaces = @interface.GetDeclaredInterfaces(basesBeingResolved);
-                        for (int i = baseInterfaces.Length - 1; i >= 0; i--)
+
+                        if (!baseInterfaces.IsEmpty)
                         {
-                            var baseInterface = baseInterfaces[i];
-                            addAllInterfaces(baseInterface, visited, result, basesBeingResolved);
+                            cycleGuard = cycleGuard.Add(@interface.OriginalDefinition);
+                            for (int i = baseInterfaces.Length - 1; i >= 0; i--)
+                            {
+                                var baseInterface = baseInterfaces[i];
+                                addAllInterfaces(baseInterface, visited, result, basesBeingResolved, cycleGuard);
+                            }
                         }
                     }
 

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs
@@ -951,7 +951,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                 return ImmutableArray<NamedTypeSymbol>.Empty;
             }
 
-            var cycleGuard = ImmutableHashSet.Create(type.OriginalDefinition);
+            var cycleGuard = ConsList<NamedTypeSymbol>.Empty.Prepend(type.OriginalDefinition);
 
             // Consumers of the result depend on the sorting performed by AllInterfacesWithDefinitionUseSiteDiagnostics.
             // Let's use similar sort algorithm.
@@ -972,17 +972,19 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             return result.ToImmutableAndFree();
 
-            static void addAllInterfaces(NamedTypeSymbol @interface, HashSet<NamedTypeSymbol> visited, ArrayBuilder<NamedTypeSymbol> result, ConsList<TypeSymbol> basesBeingResolved, ImmutableHashSet<NamedTypeSymbol> cycleGuard)
+            static void addAllInterfaces(NamedTypeSymbol @interface, HashSet<NamedTypeSymbol> visited, ArrayBuilder<NamedTypeSymbol> result, ConsList<TypeSymbol> basesBeingResolved, ConsList<NamedTypeSymbol> cycleGuard)
             {
-                if (@interface.IsInterface && !cycleGuard.Contains(@interface.OriginalDefinition) && visited.Add(@interface))
+                NamedTypeSymbol originalDefinition;
+
+                if (@interface.IsInterface && !cycleGuard.ContainsReference(originalDefinition = @interface.OriginalDefinition) && visited.Add(@interface))
                 {
-                    if (!basesBeingResolved.ContainsReference(@interface.OriginalDefinition))
+                    if (!basesBeingResolved.ContainsReference(originalDefinition))
                     {
                         ImmutableArray<NamedTypeSymbol> baseInterfaces = @interface.GetDeclaredInterfaces(basesBeingResolved);
 
                         if (!baseInterfaces.IsEmpty)
                         {
-                            cycleGuard = cycleGuard.Add(@interface.OriginalDefinition);
+                            cycleGuard = cycleGuard.Prepend(originalDefinition);
                             for (int i = baseInterfaces.Length - 1; i >= 0; i--)
                             {
                                 var baseInterface = baseInterfaces[i];

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/DefaultInterfaceImplementationTests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/DefaultInterfaceImplementationTests.cs
@@ -29513,6 +29513,37 @@ class C11
         }
 
         [Fact]
+        [WorkItem(38735, "https://github.com/dotnet/roslyn/issues/38735")]
+        public void NestedTypes_52()
+        {
+            var source1 =
+@"
+interface I1 : IA<int>.NF { }
+
+interface IQ<T> { }
+
+interface IA<T> : IB<IQ<T>> { }
+
+interface IB<T> : IA<IQ<T>> { }
+";
+
+            var compilation1 = CreateCompilation(source1, options: TestOptions.DebugDll,
+                                                 parseOptions: TestOptions.Regular);
+
+            compilation1.VerifyDiagnostics(
+                // (2,24): error CS0426: The type name 'NF' does not exist in the type 'IA<int>'
+                // interface I1 : IA<int>.NF { }
+                Diagnostic(ErrorCode.ERR_DottedTypeNameNotFoundInAgg, "NF").WithArguments("NF", "IA<int>").WithLocation(2, 24),
+                // (6,11): error CS0529: Inherited interface 'IB<IQ<T>>' causes a cycle in the interface hierarchy of 'IA<T>'
+                // interface IA<T> : IB<IQ<T>> { }
+                Diagnostic(ErrorCode.ERR_CycleInInterfaceInheritance, "IA").WithArguments("IA<T>", "IB<IQ<T>>").WithLocation(6, 11),
+                // (8,11): error CS0529: Inherited interface 'IA<IQ<T>>' causes a cycle in the interface hierarchy of 'IB<T>'
+                // interface IB<T> : IA<IQ<T>> { }
+                Diagnostic(ErrorCode.ERR_CycleInInterfaceInheritance, "IB").WithArguments("IB<T>", "IA<IQ<T>>").WithLocation(8, 11)
+                );
+        }
+
+        [Fact]
         [WorkItem(32540, "https://github.com/dotnet/roslyn/issues/32540")]
         public void MethodImplementationInDerived_01()
         {

